src/nnet3/nnet-simple-component.cc

```diff
@@ -668,12 +718,99 @@ void ClipGradientComponent::Backprop(const std::string &debug_info,
+
+  // we use this 'repair_probability' (hardcoded for now) to limit
+  // this code to running on about half of the minibatches.
+  BaseFloat repair_probability = 0.5;
+  if (self_repair_clipped_proportion_threshold_ >= 1.0 ||
+      self_repair_scale_ == 0.0 || count_ == 0 ||
+      RandUniform() > repair_probability)
+    return;
+
+  BaseFloat unset = kUnsetValue; // -1000.0
+
+  BaseFloat clipped_proportion_threshold =
+      (self_repair_clipped_proportion_threshold_ == unset ?
+       default_clipped_proportion_threshold :
+       self_repair_clipped_proportion_threshold_);
+  BaseFloat clipping_to_value =
+      (self_repair_clipping_to_value_ == unset ?
+       default_clipping_to_value :
+       self_repair_clipping_to_value_);
+  KALDI_ASSERT(clipping_to_value >= 0.0 && self_repair_scale_ > 0.0);
+
+  BaseFloat clipped_proportion =
+    (count_ > 0 ? static_cast<BaseFloat>(num_clipped_) / count_ : 0);
+  // in-deriv would be modified only when clipped_proportion exceeds the
+  // threshold
+  if (clipped_proportion <= clipped_proportion_threshold)
+    return;
+
+  to_update->num_self_repaired_ += 1;
+  if (to_update->debug_info_ == "") //debug
+    to_update->debug_info_ = debug_info; //debug
+  if (to_update->num_self_repaired_ == 1)
+    KALDI_LOG << "ClipGradientComponent(node_name=" << debug_info
+              << ")'s self-repair was activated as the first time at the "
+              << num_backproped_
+              << "-th call of Backprop() in this training job.";
+
+  // sign_mat = sign(in_value), i.e.,
+  // An element in sign_mat is 1 if its corresponding element in in_value > 0,
+  // or -1 otherwise
+  CuMatrix<BaseFloat> sign_mat(in_value);
+  sign_mat.ApplyHeaviside();
+  sign_mat.Scale(2.0);
+  sign_mat.Add(-1.0);
+
+  // repair_mat =
+  // floor(abs(in_value) - clipping_to_value, 0) .* sign(in_value)
+  CuMatrix<BaseFloat> repair_mat(in_value);
+  repair_mat.ApplyPowAbs(1.0);
+  repair_mat.Add(-clipping_to_value);
+  repair_mat.ApplyFloor(0.0);
+  repair_mat.MulElements(sign_mat);
+
+  // magnitude =
+  // self_repair_scale_ * clipped_proportion * average norm of in-deriv
+  CuVector<BaseFloat> in_deriv_norm_vec(in_deriv->NumRows());
+  in_deriv_norm_vec.AddDiagMat2(1.0, *in_deriv, kNoTrans, 0.0);
+  in_deriv_norm_vec.ApplyPow(0.5);
+  double in_deriv_norm_sum = in_deriv_norm_vec.Sum();
+  BaseFloat magnitude = self_repair_scale_ * clipped_proportion *
+                        (in_deriv_norm_sum / in_deriv_norm_vec.Dim());
+ 
+  CuVector<BaseFloat> scales(repair_mat.NumRows());
+  scales.AddDiagMat2(pow(magnitude, -2), repair_mat, kNoTrans, 0.0);
+  // here we call ApplyPowAbs() instead of ApplyPow() because
+  // we want the inverse of 0 to be 0, rather than NaN(or whatever else).
+  scales.ApplyPowAbs(-0.5);
+  // now each element in scales = magnitude / (norm of each row of repair_mat)
+
+  // each row of the matrix being added to in_deriv is scaled to have the norm
+  // (magnitude / repair_probability).
```

Comment:
Please verify with a print statement that each row of the matrix really does have the norm (magnitude / repair_probability)... and then remove the print statement.  There could be a bug somewhere.